File: front/src/type/quiz.ts
export type Options = 1 | 2 | 3 | 4;

export interface Quiz {
  id: string;
  correct_option: Options;
  created_by: string;
  option1: string;
  option2: string;
  option3: string;
  option4: string;
  question: string;
  created_at?: Date;
  updated_at?: Date;
}

export interface GetQuizzesRes {
  data: Quiz[];
}

export interface PostInput {
  correctOption: Options;
  createdBy: string;
  option1: string;
  option2: string;
  option3: string;
  option4: string;
  question: string;
}

export interface EditInput {
  id: string;
  correctOption: Options;
  option1: string;
  option2: string;
  option3: string;
  option4: string;
  question: string;
}

export interface DeleteInput {
  id: string;
}

File: api/src/models/quiz.rs
use super::user::User;
use chrono::NaiveDateTime;
use diesel::prelude::*;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Queryable, Identifiable, Associations, Serialize, Deserialize, Clone)]
#[diesel(table_name = crate::schema::quizzes)]
#[diesel(check_for_backend(diesel::pg::Pg))]
#[diesel(belongs_to(User, foreign_key = created_by))]
pub struct Quiz {
    pub id: Uuid,
    pub correct_option: i32,
    pub created_by: Uuid,
    pub option1: String,
    pub option2: String,
    pub option3: String,
    pub option4: String,
    pub question: String,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
}

#[derive(Insertable, Serialize, Deserialize)]
#[diesel(table_name = crate::schema::quizzes)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct NewQuiz {
    pub correct_option: i32,
    pub created_by: Uuid,
    pub option1: String,
    pub option2: String,
    pub option3: String,
    pub option4: String,
    pub question: String,
}

#[derive(AsChangeset, Identifiable, Serialize, Deserialize)]
#[diesel(table_name = crate::schema::quizzes)]
#[diesel(check_for_backend(diesel::pg::Pg))]

pub struct UpdateQuiz {
    pub id: Uuid,
    pub correct_option: i32,
    pub option1: String,
    pub option2: String,
    pub option3: String,
    pub option4: String,
    pub question: String,
}
